### PR TITLE
refactor(java_binding): extract java_binding to jni_core

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Related tracking issue https://github.com/risingwavelabs/risingwave/issues/12090
- Extract `java_binding` into `jni_core`, so that we can share common logic to register native methods via `jni-rs` or compile a dynamic library in the following PRs.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/java_binding/src/lib.rs [hunk 1]
```diff
@@ -17,821 +17,6 @@
 #![feature(lazy_cell)]
 #![feature(once_cell_try)]
 #![feature(type_alias_impl_trait)]
+#![feature(result_option_inspect)]
```
Comment (wenym1) [inline src/java_binding/src/lib.rs:20]: What is this feature for?

File: src/java_binding/Cargo.toml [hunk 2]
```diff
@@ -17,6 +17,7 @@ jni = "0.21.1"
 prost = "0.11"
 risingwave_common = { workspace = true }
 risingwave_hummock_sdk = { workspace = true }
+risingwave_jni_core = { workspace = true }
```
Comment (wenym1) [inline src/java_binding/Cargo.toml:20]: I guess here we only depends on `risingwave_jni_core`, and we can remove the rest.

File: src/meta/Cargo.toml [hunk 3]
```diff
@@ -49,6 +49,7 @@ risingwave_common = { workspace = true }
 risingwave_common_service = { workspace = true }
 risingwave_connector = { workspace = true }
 risingwave_hummock_sdk = { workspace = true }
+risingwave_jni_core = { workspace = true }
```
Comment (wenym1) [inline src/meta/Cargo.toml:52]: In this PR, this dependency may not be included yet?

**Review by wenym1 — APPROVED** [review]

**Review by fuyufjh — APPROVED** [review]
mostly code structure change

**Review by BugenZhao — APPROVED** [review]

**Review by StrikeW — APPROVED** [review]

**Pull request merged**